Extensions/ArtifactEngine/Engine/worker.ts

```diff
@@ -1,52 +1,53 @@
 import { Logger } from './logger';
 export class Worker<T> {
-    private execute: (item : T) => Promise<void>;
+    private execute: (item: T) => Promise<void>;
     private getNextItem: () => T;
     private canExit: () => boolean;
 
     private id: number;
 
-    constructor(id: number, execute: (item : T) => Promise<void>, getNextItem: () => T, canExit: () => boolean) {
+    constructor(id: number, execute: (item: T) => Promise<void>, getNextItem: () => T, canExit: () => boolean) {
         this.id = id;
         this.execute = execute;
         this.getNextItem = getNextItem;
         this.canExit = canExit;
     }
 
     init(): Promise<void> {
-        var promise = new Promise<void>(async (resolve, reject) => {
-            await this.spawnWorker(resolve, reject);
+        var promise = new Promise<void>((resolve, reject) => {
+            this.spawnWorker(resolve, reject);
         });
 
         return promise;
     }
 
-    async spawnWorker(resolve, reject) {
+    spawnWorker(resolve, reject) {
         try {
-            await this.workerImplementation();
-            if (this.canExit()) {
-                Logger.logInfo(`Exiting worker ${this.id} nothing more to process`);
-                resolve();
-            }
-            else {
-                Logger.logInfo(`Spawn worker ${this.id} after 1 sec to check for items again.`);
+            let item = this.getNextItem();
+            if (!item && !this.canExit()) {
+                Logger.logInfo(`Nothing to process currently, respawing worker ${this.id} after 1 sec.`);
                 setTimeout(() => this.spawnWorker(resolve, reject), 1000);
+                return;
             }
-        }
-        catch (err) {
-            reject(err);
-        }
-    }
-    
-    async workerImplementation() {
-        while (true) {
-            const item = this.getNextItem();
+
             if (!item) {
-                Logger.logInfo(`Nothing more to process, exiting workerImplementation for worker ${this.id}.`);
-                break;
+                Logger.logInfo(`Nothing more to process, exiting worker ${this.id}.`);
+                resolve();
+                return;
             }
 
-            await this.execute(item);
+            let executePromise = this.execute(item);
+
+            executePromise.then(() => {
+                Logger.logInfo(`Nothing to process currently, respawing worker ${this.id} after 1 sec.`);
+                this.spawnWorker(resolve, reject);
+            }, (reason) => {
+                reject(reason);
+                return;
+            });
+        }
+        catch (err) {
```

Comment:
what do we expect to catch here?

Reply:
exception isn't expected here, but if anything goes wrong, worker should reject.